app/services/coffee_shop_service.rb

```diff
@@ -0,0 +1,32 @@
+# frozen_string_literal: true
+
+module Services
+  class CoffeeShopService
+    include Helpers::DistanceHelper
+
+    def initialize(user_coords, coffee_shops)
+      @user_coords = user_coords
+      @coffee_shops = coffee_shops
+    end
+
+    def call
+      closest_shops(calculate_distances_hash)
+    end
+
+    private
+
+    def calculate_distances_hash
+      @coffee_shops.map do |shop|
+        distance = calculate_euclidean_distance(
+          @user_coords[:x].to_f, @user_coords[:y].to_f,
+          shop[:x], shop[:y]
+        )
+        shop.merge(distance: distance)
+      end
+    end
+
+    def closest_shops(shops)
+      shops.sort_by { |shop| shop[:distance] }.first(3)
+    end
```

Comment:
This can be optimized a little bit

Reply:
I can rewrite the whole service like this
`
module Services
  class CoffeeShopService
    include Helpers::DistanceHelper

    def initialize(user_coords, coffee_shops)
      @user_coords = user_coords
      @coffee_shops = coffee_shops
    end

    def call
      closest_shop
    end

    private

    def closest_shop
      user_x, user_y = @user_coords.values_at(:x, :y).map(&:to_f)
      
      @coffee_shops.map do |shop|
        shop.merge(distance: calculate_euclidean_distance(user_x, user_y, shop[:x], shop[:y]))
      end.min_by(3) { |shop| shop[:distance] }
    end
    
  end
end
`

sort is O(n log n) and min_by is O(n).